### PR TITLE
Extend password reset token validity on password change page load

### LINKED ISSUE #29647: [Issue] Extend password reset token validity on password change page load
This issue is automatically created based on existing pull request: magento/magento2#25279: Extend password reset token validity on password change page load

---------

### Description (*)

This avoids a race condition where the password reset token is valid at page load time, but no longer valid at page submission time.

In a scenario where a user begins the password reset flow but is unable to immediately complete the same, clicking the link close to expiry time provides an opportunity for an error message after entering a new password due to the token expiring between page load & form submission.

### Manual testing scenarios (*)

1. Generate password reset email (admin & frontend)
1. Follow link back to Magento
1. Observe password validity in database has been extended/reset

### Questions or comments

I am unsure if this change requires a unit test.

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by fredden**
### Description (*)

This avoids a race condition where the password reset token is valid at page load time, but no longer valid at page submission time.

In a scenario where a user begins the password reset flow but is unable to immediately complete the same, clicking the link close to expiry time provides an opportunity for an error message after entering a new password due to the token expiring between page load & form submission.

### Manual testing scenarios (*)

1. Generate password reset email (admin & frontend)
1. Follow link back to Magento
1. Observe password validity in database has been extended/reset

### Questions or comments

I am unsure if this change requires a unit test.

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#29647: Extend password reset token validity on password change page load

### CODE REVIEW

**Comment by sidolov** [comment]
@magento create issue

**Comment by ihor-sviziev** [comment]
Hi @fredden,
I'm really sorry for long delay with processing your PR. I'll take it and will try to push it forward.

**Comment by ihor-sviziev** [comment]
@magento run all tests

File: app/code/Magento/Customer/Controller/Account/CreatePassword.php [hunk 1]
```diff
@@ -83,6 +91,15 @@ public function execute()
 
             $this->confirmByToken->execute($resetPasswordToken);
 
+            try {
+                // Extend token validity to avoid expiration while this form is
+                // being completed by the user.
+                $customer = $this->getByToken->execute($resetPasswordToken);
+                $this->accountManagement->changeResetPasswordLinkToken($customer, $resetPasswordToken);
+            } catch (\Exception $exception) {
+                // Intentionally ignoring failures here
```
Comment (ihor-sviziev) [inline app/code/Magento/Customer/Controller/Account/CreatePassword.php]: 1. Catching all types exceptions isn't really good thing. I believe we shouldn't catch PDOException here, for instance? Can we catch there only exceptions that we really expect?

2. Empty catch block also not a good thing, usually it just masks the problems. Which failures we expect here? Why should we ignore them? Should we add some message or maybe log them in some case?

File: app/code/Magento/User/Controller/Adminhtml/Auth/ResetPassword.php [hunk 2]
```diff
@@ -22,6 +26,16 @@ public function execute()
...
             $this->_validateResetPasswordLinkToken($userId, $passwordResetToken);
 
+            try {
+                // Extend token validity to avoid expiration while this form is
+                // being completed by the user.
+                $user = $this->_userFactory->create()->load($userId);
+                $user->changeResetPasswordLinkToken($passwordResetToken);
+                $user->save();
+            } catch (\Exception $exception) {
+                // Intentionally ignoring failures here
```
Comment (ihor-sviziev) [inline app/code/Magento/User/Controller/Adminhtml/Auth/ResetPassword.php]: Basically the same here as for prev. catch block.

File: app/code/Magento/Customer/Controller/Account/CreatePassword.php [hunk 3]
```diff
@@ -42,25 +41,34 @@ class CreatePassword extends \Magento\Customer\Controller\AbstractAccount implem
      */
     private $confirmByToken;
 
+    /**
+     * @var \Magento\Customer\Model\ForgotPasswordToken\GetCustomerByToken
```
Comment (ihor-sviziev) [inline app/code/Magento/Customer/Controller/Account/CreatePassword.php]: Looks like this class already imported, so we can change it to something like this
```suggestion
     * @var GetCustomerByToken
```

File: app/code/Magento/Customer/Controller/Account/CreatePassword.php [hunk 4]
```diff
@@ -42,25 +41,34 @@ class CreatePassword extends \Magento\Customer\Controller\AbstractAccount implem
...
+     */
+    private $getByToken;
+
     /**
      * @param \Magento\Framework\App\Action\Context $context
      * @param \Magento\Customer\Model\Session $customerSession
      * @param \Magento\Framework\View\Result\PageFactory $resultPageFactory
      * @param \Magento\Customer\Api\AccountManagementInterface $accountManagement
      * @param \Magento\Customer\Model\ForgotPasswordToken\ConfirmCustomerByToken $confirmByToken
+     * @param \Magento\Customer\Model\ForgotPasswordToken\GetCustomerByToken $getByToken
```
Comment (ihor-sviziev) [inline app/code/Magento/Customer/Controller/Account/CreatePassword.php]: Seems like all listed arguments are already imported, and some of them are not always an object, so we can update it to this
```suggestion
     * @param Context $context
     * @param Session $customerSession
     * @param PageFactory $resultPageFactory
     * @param AccountManagementInterface $accountManagement
     * @param ConfirmCustomerByToken|null $confirmByToken
     * @param GetCustomerByToken|null $getByToken
```

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Comment by engcom-Charlie** [comment]
@fredden can you please look at the requested changes and cover your changes by some kind of autotests? Otherwise, we can't proceed with your PR.
Thank you.

**Comment by sidolov** [comment]
@fredden  I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for the collaboration!

**Comment by fredden** [comment]
Please may I have some guidance on how to write good tests to cover this change?

**Comment by fredden** [comment]
@sidolov I don't have a 'reopen' button; please can you re-open this.

**Comment by ihor-sviziev** [comment]
@fredden I just reopened this PR by your request. Please update PR by the requested earlier changes.
From the tests perspective - I think the most useful will be here integration tests

**Comment by ihor-sviziev** [comment]
@sidolov could you explain why this PR was assigend to GraphQL board?

**Comment by sidolov** [comment]
@ihor-sviziev it's caused by the issue in the app, it's successfully resolved and PRs will be assigned to the correct board

**Comment by engcom-Charlie** [comment]
@fredden please update PR accordingly to the requested earlier changes and cover your changes by integration test.
Thank you.

**Comment by engcom-Charlie** [comment]
@fredden could you please cover your changes by integration test?
Thank you.

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by fredden** [comment]
@magento run WebAPI Tests

**Comment by fredden** [comment]
@magento run all tests

**Comment by fredden** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
Hi @fredden,
Thank you for covering your changes with integration tests! Could you fix the failing static test (just ignore the coupling between objects in auth test file) and I’ll approve your PR?

**Comment by fredden** [comment]
@ihor-sviziev the only complaint from the automated testing system is the one you're suggesting I ignore.

> PHP Code Mess has found error(s):
/var/www/html/dev/tests/integration/testsuite/Magento/User/Controller/Adminhtml/AuthTest.php:18	The class AuthTest has a coupling between objects value of 15. Consider to reduce the number of dependencies under 13.

How do you suggest we proceed?

**Comment by ihor-sviziev** [comment]
@fredden please add such line `@SuppressWarnings(PHPMD.CouplingBetweenObjects)` to the phpdoc block for the classs, example:
https://github.com/magento/magento2/blob/9544fb243d5848a497d4ea7b88e08609376ac39e/dev/tests/integration/framework/Magento/TestFramework/Catalog/Model/Indexer/Category/Product/Action/Full.php#L13-L19

**Comment by fredden** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests EE, WebAPI Tests

**Comment by fredden** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B, Functional Tests CE

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8461](https://jira.corp.magento.com/browse/ENGCOM-8461) has been created to process this Pull Request

**Comment by engcom-Delta** [comment]
:heavy_check_mark: QA passed
**Before:**
:x: Value in `rp_token_created_at` column is not changed after customer open a `Set a New Password` link from reset password email

**After:**
:heavy_check_mark: Value in `rp_token_created_at` column is updated and is equal to the time when customer open a `Set a New Password` link from reset password email
Peek 2020-11-26 14-36

**Comment by engcom-Delta** [comment]
**Note:** Automation tests are passed

**Comment by korostii** [comment]
@magento-engcom-team Any progress update here please?

**Pull request merged**